crates/oxc_linter/src/rules/unicorn/consistent_function_scoping.rs

```diff
@@ -0,0 +1,1233 @@
+                    )
+                } else {
+                    return;
+                }
+            }
+            _ => return,
+        };
+        // if the function is declared at the root scope, we don't need to check anything
+        if ctx.symbols().get_scope_id(function_declaration_symbol_id)
+            == ctx.scopes().root_scope_id()
+        {
+            return;
+        }
+
+        if let Some(AstKind::ReturnStatement(_)) =
+            outermost_paren_parent(node, ctx).map(oxc_semantic::AstNode::kind)
+        {
+            return;
+        }
+
+        if is_parent_scope_iife(node, ctx) || is_in_react_hook(node, ctx) {
+            return;
+        }
+
+        // get all references in the function body
+        let (function_body_var_references, is_parent_this_referenced) = {
+            let mut rf = ReferencesFinder::default();
+            rf.visit_function_body(function_body);
+            (rf.references, rf.is_parent_this_referenced)
+        };
+
+        if is_parent_this_referenced && matches!(node.kind(), AstKind::ArrowFunctionExpression(_)) {
+            return;
+        }
+
+        let parent_scope_ids = {
+            let mut current_scope_id = ctx.symbols().get_scope_id(function_declaration_symbol_id);
+            let mut parent_scope_ids = FxHashSet::default();
+            parent_scope_ids.insert(current_scope_id);
+            while let Some(parent_scope_id) = ctx.scopes().get_parent_id(current_scope_id) {
+                parent_scope_ids.insert(parent_scope_id);
+                current_scope_id = parent_scope_id;
+            }
+            parent_scope_ids
+        };
+
+        for reference_id in function_body_var_references {
+            let reference = ctx.symbols().get_reference(reference_id);
+            let Some(symbol_id) = reference.symbol_id() else { continue };
+            let scope_id = ctx.symbols().get_scope_id(symbol_id);
+            if parent_scope_ids.contains(&scope_id) && symbol_id != function_declaration_symbol_id {
+                return;
+            }
+        }
+
+        ctx.diagnostic(
+            OxcDiagnostic::warn("Function does not capture any variables from the outer scope.")
+                .with_help("Move this function to the outer scope.")
+                .with_label(reporter_span),
+        );
+    }
+}
+
+#[derive(Default)]
+struct ReferencesFinder {
+    is_parent_this_referenced: bool,
+    references: Vec<ReferenceId>,
+    in_function: usize,
+}
+
+impl<'a> Visit<'a> for ReferencesFinder {
+    fn visit_identifier_reference(&mut self, it: &oxc_ast::ast::IdentifierReference<'a>) {
+        self.references.push(it.reference_id().unwrap());
+    }
+
+    fn visit_this_expression(&mut self, _: &oxc_ast::ast::ThisExpression) {
+        if self.in_function == 0 {
+            self.is_parent_this_referenced = true;
+        }
+    }
+
+    fn enter_node(&mut self, kind: AstKind<'a>) {
+        if let AstKind::Function(_) = kind {
```

Comment:
What about arrow functions?

Reply:
i think... because arrow funcitons inherit the `this` from the parent scope whereas `function foo() {}` has it's own `this` value. therefore, it's safe to move `function foo() { this }` to the outer scope, but not `() => this`